// @ionic-native/photo-library/index.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
import { Plugin, Cordova, IonicNativePlugin, wrap } from '@ionic-native/core';
import { Observable } from 'rxjs/Observable';
import { Injectable } from '@angular/core';
/**
 * @hidden
 *
 * Wraps method that returns an observable that can be completed. Provided opts.resultFinalPredicate dictates when the observable completes.
 */
export function CordovaFiniteObservable(opts) {
    if (opts === void 0) { opts = {}; }
    opts.observable = true;
    return function (target, methodName, descriptor) {
        return {
            value: function () {
                var args = [];
                for (var _i = 0; _i < arguments.length; _i++) {
                    args[_i] = arguments[_i];
                }
                var wrappedObservable = wrap(this, methodName, opts).apply(this, args);
                return new Observable(function (observer) {
                    var wrappedSubscription = wrappedObservable.subscribe({
                        next: function (x) {
                            observer.next(opts.resultTransform ? x[opts.resultTransform] : x);
                            if (opts.resultFinalPredicate && x[opts.resultFinalPredicate]) {
                                observer.complete();
                            }
                        },
                        error: function (err) { observer.error(err); },
                        complete: function () { observer.complete(); }
                    });
                    return function () {
                        wrappedSubscription.unsubscribe();
                    };
                });
            },
            enumerable: true
        };
    };
}
/**
 * @name Photo Library
 * @description
 * The PhotoLibrary plugin allows access to photos from device by url. So you can use plain img tag to display photos and their thumbnails, and different 3rd party libraries as well.
 * Saving photos and videos to the library is also supported.
 * cdvphotolibrary urls should be trusted by Angular. See plugin homepage to learn how.
 *
 * @usage
 * ```typescript
 * import { PhotoLibrary } from '@ionic-native/photo-library';
 *
 * constructor(private photoLibrary: PhotoLibrary) { }
 *
 * this.photoLibrary.requestAuthorization().then(() => {
 *   this.photoLibrary.getLibrary().subscribe({
 *     next: library => {
 *       library.forEach(function(libraryItem) {
 *         console.log(libraryItem.id);          // ID of the photo
 *         console.log(libraryItem.photoURL);    // Cross-platform access to photo
 *         console.log(libraryItem.thumbnailURL);// Cross-platform access to thumbnail
 *         console.log(libraryItem.fileName);
 *         console.log(libraryItem.width);
 *         console.log(libraryItem.height);
 *         console.log(libraryItem.creationDate);
 *         console.log(libraryItem.latitude);
 *         console.log(libraryItem.longitude);
 *         console.log(libraryItem.albumIds);    // array of ids of appropriate AlbumItem, only of includeAlbumsData was used
 *       });
 *     },
 *     error: err => { console.log('could not get photos'); },
 *     complete: () => { console.log('done getting photos'); }
 *   });
 * })
 * .catch(err => console.log('permissions weren\'t granted'));
 *
 * ```
 */
var PhotoLibrary = (function (_super) {
    __extends(PhotoLibrary, _super);
    function PhotoLibrary() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    /**
     * Retrieves library items. Library item contains photo metadata like width and height, as well as photoURL and thumbnailURL.
     * @param options {GetLibraryOptions} Optional, like thumbnail size and chunks settings.
     * @return {Observable<LibraryItem[]>} Returns library items. If appropriate option was set, will be returned by chunks.
     */
    /**
       * Retrieves library items. Library item contains photo metadata like width and height, as well as photoURL and thumbnailURL.
       * @param options {GetLibraryOptions} Optional, like thumbnail size and chunks settings.
       * @return {Observable<LibraryItem[]>} Returns library items. If appropriate option was set, will be returned by chunks.
       */
    PhotoLibrary.prototype.getLibrary = /**
       * Retrieves library items. Library item contains photo metadata like width and height, as well as photoURL and thumbnailURL.
       * @param options {GetLibraryOptions} Optional, like thumbnail size and chunks settings.
       * @return {Observable<LibraryItem[]>} Returns library items. If appropriate option was set, will be returned by chunks.
       */
    function (options) { return; };
    /**
     * Asks user permission to access photo library.
     * @param options {RequestAuthorizationOptions} Optional, like whether only read access needed or read/write.
     * @return { Promise<void>} Returns a promise that resolves when permissions are granted, and fails when not.
     */
    /**
       * Asks user permission to access photo library.
       * @param options {RequestAuthorizationOptions} Optional, like whether only read access needed or read/write.
       * @return { Promise<void>} Returns a promise that resolves when permissions are granted, and fails when not.
       */
    PhotoLibrary.prototype.requestAuthorization = /**
       * Asks user permission to access photo library.
       * @param options {RequestAuthorizationOptions} Optional, like whether only read access needed or read/write.
       * @return { Promise<void>} Returns a promise that resolves when permissions are granted, and fails when not.
       */
    function (options) { return; };
    /**
     * Returns list of photo albums on device.
     * @return {Promise<AlbumItem[]>} Resolves to list of albums.
     */
    /**
       * Returns list of photo albums on device.
       * @return {Promise<AlbumItem[]>} Resolves to list of albums.
       */
    PhotoLibrary.prototype.getAlbums = /**
       * Returns list of photo albums on device.
       * @return {Promise<AlbumItem[]>} Resolves to list of albums.
       */
    function () { return; };
    /**
     * Provides means to request URL of thumbnail, with specified size or quality.
     * @param photo {string | LibraryItem} Id of photo, or LibraryItem.
     * @param options {GetThumbnailOptions} Options, like thumbnail size or quality.
     * @return {Promise<string>} Resolves to URL of cdvphotolibrary schema.
     */
    /**
       * Provides means to request URL of thumbnail, with specified size or quality.
       * @param photo {string | LibraryItem} Id of photo, or LibraryItem.
       * @param options {GetThumbnailOptions} Options, like thumbnail size or quality.
       * @return {Promise<string>} Resolves to URL of cdvphotolibrary schema.
       */
    PhotoLibrary.prototype.getThumbnailURL = /**
       * Provides means to request URL of thumbnail, with specified size or quality.
       * @param photo {string | LibraryItem} Id of photo, or LibraryItem.
       * @param options {GetThumbnailOptions} Options, like thumbnail size or quality.
       * @return {Promise<string>} Resolves to URL of cdvphotolibrary schema.
       */
    function (photo, options) { return; };
    /**
     * Provides means to request photo URL by id.
     * @param photo {string | LibraryItem} Id or LibraryItem.
     * @param options {GetPhotoOptions} Optional options.
     * @return {Promise<string>} Resolves to URL of cdvphotolibrary schema.
     */
    /**
       * Provides means to request photo URL by id.
       * @param photo {string | LibraryItem} Id or LibraryItem.
       * @param options {GetPhotoOptions} Optional options.
       * @return {Promise<string>} Resolves to URL of cdvphotolibrary schema.
       */
    PhotoLibrary.prototype.getPhotoURL = /**
       * Provides means to request photo URL by id.
       * @param photo {string | LibraryItem} Id or LibraryItem.
       * @param options {GetPhotoOptions} Optional options.
       * @return {Promise<string>} Resolves to URL of cdvphotolibrary schema.
       */
    function (photo, options) { return; };
    /**
     * Returns thumbnail as Blob.
     * @param photo {string | LibraryItem} Id or LibraryItem.
     * @param options {GetThumbnailOptions} Options, like thumbnail size or quality.
     * @return {Promise<Blob>} Resolves requested thumbnail as blob.
     */
    /**
       * Returns thumbnail as Blob.
       * @param photo {string | LibraryItem} Id or LibraryItem.
       * @param options {GetThumbnailOptions} Options, like thumbnail size or quality.
       * @return {Promise<Blob>} Resolves requested thumbnail as blob.
       */
    PhotoLibrary.prototype.getThumbnail = /**
       * Returns thumbnail as Blob.
       * @param photo {string | LibraryItem} Id or LibraryItem.
       * @param options {GetThumbnailOptions} Options, like thumbnail size or quality.
       * @return {Promise<Blob>} Resolves requested thumbnail as blob.
       */
    function (photo, options) { return; };
    /**
     * Returns photo as Blob.
     * @param photo {string | LibraryItem} Id or LibraryItem.
     * @param options {GetPhotoOptions} Optional options.
     * @return {Promise<Blob>} Resolves requested photo as blob.
     */
    /**
       * Returns photo as Blob.
       * @param photo {string | LibraryItem} Id or LibraryItem.
       * @param options {GetPhotoOptions} Optional options.
       * @return {Promise<Blob>} Resolves requested photo as blob.
       */
    PhotoLibrary.prototype.getPhoto = /**
       * Returns photo as Blob.
       * @param photo {string | LibraryItem} Id or LibraryItem.
       * @param options {GetPhotoOptions} Optional options.
       * @return {Promise<Blob>} Resolves requested photo as blob.
       */
    function (photo, options) { return; };
    /**
     * Saves image to specified album. Album will be created if not exists.
     * LibraryItem that represents saved image is returned.
     * @param url {string} URL of a file, or DataURL.
     * @param album {AlbumItem | string} Name of an album or AlbumItem object.
     * @param options {GetThumbnailOptions} Options, like thumbnail size for resulting LibraryItem.
     * @return {Promise<LibraryItem>} Resolves to LibraryItem that represents saved image.
     */
    /**
       * Saves image to specified album. Album will be created if not exists.
       * LibraryItem that represents saved image is returned.
       * @param url {string} URL of a file, or DataURL.
       * @param album {AlbumItem | string} Name of an album or AlbumItem object.
       * @param options {GetThumbnailOptions} Options, like thumbnail size for resulting LibraryItem.
       * @return {Promise<LibraryItem>} Resolves to LibraryItem that represents saved image.
       */
    PhotoLibrary.prototype.saveImage = /**
       * Saves image to specified album. Album will be created if not exists.
       * LibraryItem that represents saved image is returned.
       * @param url {string} URL of a file, or DataURL.
       * @param album {AlbumItem | string} Name of an album or AlbumItem object.
       * @param options {GetThumbnailOptions} Options, like thumbnail size for resulting LibraryItem.
       * @return {Promise<LibraryItem>} Resolves to LibraryItem that represents saved image.
       */
    function (url, album, options) { return; };
    /**
     * Saves video to specified album. Album will be created if not exists.
     * @param url {string} URL of a file, or DataURL.
     * @param album {AlbumItem | string} Name of an album or AlbumItem object.
     * @return {Promise<void>} Resolves when save operation completes.
     */
    /**
       * Saves video to specified album. Album will be created if not exists.
       * @param url {string} URL of a file, or DataURL.
       * @param album {AlbumItem | string} Name of an album or AlbumItem object.
       * @return {Promise<void>} Resolves when save operation completes.
       */
    PhotoLibrary.prototype.saveVideo = /**
       * Saves video to specified album. Album will be created if not exists.
       * @param url {string} URL of a file, or DataURL.
       * @param album {AlbumItem | string} Name of an album or AlbumItem object.
       * @return {Promise<void>} Resolves when save operation completes.
       */
    function (url, album) { return; };
    PhotoLibrary.decorators = [
        { type: Injectable },
    ];
    __decorate([
        CordovaFiniteObservable({
            callbackOrder: 'reverse',
            resultFinalPredicate: 'isLastChunk',
            resultTransform: 'library'
        }),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", [Object]),
        __metadata("design:returntype", Observable)
    ], PhotoLibrary.prototype, "getLibrary", null);
    __decorate([
        Cordova({
            callbackOrder: 'reverse'
        }),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", [Object]),
        __metadata("design:returntype", Promise)
    ], PhotoLibrary.prototype, "requestAuthorization", null);
    __decorate([
        Cordova({
            callbackOrder: 'reverse'
        }),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", []),
        __metadata("design:returntype", Promise)
    ], PhotoLibrary.prototype, "getAlbums", null);
    __decorate([
        Cordova({
            successIndex: 1,
            errorIndex: 2
        }),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", [Object, Object]),
        __metadata("design:returntype", Promise)
    ], PhotoLibrary.prototype, "getThumbnailURL", null);
    __decorate([
        Cordova({
            successIndex: 1,
            errorIndex: 2
        }),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", [Object, Object]),
        __metadata("design:returntype", Promise)
    ], PhotoLibrary.prototype, "getPhotoURL", null);
    __decorate([
        Cordova({
            successIndex: 1,
            errorIndex: 2
        }),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", [Object, Object]),
        __metadata("design:returntype", Promise)
    ], PhotoLibrary.prototype, "getThumbnail", null);
    __decorate([
        Cordova({
            successIndex: 1,
            errorIndex: 2
        }),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", [Object, Object]),
        __metadata("design:returntype", Promise)
    ], PhotoLibrary.prototype, "getPhoto", null);
    __decorate([
        Cordova({
            successIndex: 2,
            errorIndex: 3
        }),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", [String, Object, Object]),
        __metadata("design:returntype", Promise)
    ], PhotoLibrary.prototype, "saveImage", null);
    __decorate([
        Cordova({
            successIndex: 2,
            errorIndex: 3
        }),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", [String, Object]),
        __metadata("design:returntype", Promise)
    ], PhotoLibrary.prototype, "saveVideo", null);
    /**
     * @name Photo Library
     * @description
     * The PhotoLibrary plugin allows access to photos from device by url. So you can use plain img tag to display photos and their thumbnails, and different 3rd party libraries as well.
     * Saving photos and videos to the library is also supported.
     * cdvphotolibrary urls should be trusted by Angular. See plugin homepage to learn how.
     *
     * @usage
     * ```typescript
     * import { PhotoLibrary } from '@ionic-native/photo-library';
     *
     * constructor(private photoLibrary: PhotoLibrary) { }
     *
     * this.photoLibrary.requestAuthorization().then(() => {
     *   this.photoLibrary.getLibrary().subscribe({
     *     next: library => {
     *       library.forEach(function(libraryItem) {
     *         console.log(libraryItem.id);          // ID of the photo
     *         console.log(libraryItem.photoURL);    // Cross-platform access to photo
     *         console.log(libraryItem.thumbnailURL);// Cross-platform access to thumbnail
     *         console.log(libraryItem.fileName);
     *         console.log(libraryItem.width);
     *         console.log(libraryItem.height);
     *         console.log(libraryItem.creationDate);
     *         console.log(libraryItem.latitude);
     *         console.log(libraryItem.longitude);
     *         console.log(libraryItem.albumIds);    // array of ids of appropriate AlbumItem, only of includeAlbumsData was used
     *       });
     *     },
     *     error: err => { console.log('could not get photos'); },
     *     complete: () => { console.log('done getting photos'); }
     *   });
     * })
     * .catch(err => console.log('permissions weren\'t granted'));
     *
     * ```
     */
    PhotoLibrary = __decorate([
        Plugin({
            pluginName: 'PhotoLibrary',
            plugin: 'cordova-plugin-photo-library',
            pluginRef: 'cordova.plugins.photoLibrary',
            repo: 'https://github.com/terikon/cordova-plugin-photo-library',
            install: 'ionic cordova plugin add cordova-plugin-photo-library --variable PHOTO_LIBRARY_USAGE_DESCRIPTION="To choose photos"',
            installVariables: ['PHOTO_LIBRARY_USAGE_DESCRIPTION'],
            platforms: ['Android', 'Browser', 'iOS']
        })
    ], PhotoLibrary);
    return PhotoLibrary;
}(IonicNativePlugin));
export { PhotoLibrary };
